core/git.go

```diff
@@ -111,3 +274,154 @@ func (ref *GitRef) getState(ctx context.Context, discardGitDir bool) (llb.State,
+	tempFile, err := os.CreateTemp("", "known_hosts")
+	if err != nil {
+		return "", fmt.Errorf("failed to create temporary known_hosts file: %w", err)
+	}
+
+	_, err = tempFile.WriteString(knownHosts)
+	if err != nil {
+		os.Remove(tempFile.Name())
+		return "", fmt.Errorf("failed to write known_hosts content: %w", err)
+	}
+
+	err = tempFile.Close()
+	if err != nil {
+		os.Remove(tempFile.Name())
+		return "", fmt.Errorf("failed to close temporary known_hosts file: %w", err)
+	}
+
+	return tempFile.Name(), nil
+}
+
+type LocalGitRepository struct {
+	Query *Query
+
+	Directory *Directory
+}
+
+var _ GitRepositoryBackend = (*LocalGitRepository)(nil)
+
+func (repo *LocalGitRepository) Head(ctx context.Context) (GitRefBackend, error) {
+	return &LocalGitRef{
+		Query: repo.Query,
+		Repo:  repo,
+		Ref:   "HEAD",
+	}, nil
+}
+
+func (repo *LocalGitRepository) Ref(ctx context.Context, ref string) (GitRefBackend, error) {
+	return &LocalGitRef{
+		Query: repo.Query,
+		Repo:  repo,
+		Ref:   ref,
+	}, nil
+}
+
+func (repo *LocalGitRepository) Tags(ctx context.Context, patterns []string) ([]string, error) {
+	var tags []string
+	err := repo.mount(ctx, func(src string) error {
+		out, err := gitCmd(ctx, src, "tag", "-l")
+		if err != nil {
+			return err
+		}
+		tags = strings.Split(out, "\n")
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return tags, nil
+}
+
+func (repo *LocalGitRepository) mount(ctx context.Context, f func(string) error) error {
+	return repo.Directory.mount(ctx, func(root string) error {
+		src, err := fs.RootPath(root, repo.Directory.Dir)
+		if err != nil {
+			return err
+		}
+		return f(src)
+	})
+}
+
+type LocalGitRef struct {
+	Query *Query
+
+	// XXX: error handling if git repo does not exist
+	Repo *LocalGitRepository
+	Ref  string
+}
+
+var _ GitRefBackend = (*LocalGitRef)(nil)
+
+func (ref *LocalGitRef) Tree(ctx context.Context, discardGitDir bool) (*Directory, error) {
+	tmpDir, err := os.MkdirTemp("", "local-git-checkout")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	err = ref.Repo.mount(ctx, func(src string) error {
+		if _, err := gitCmd(ctx, tmpDir, "init"); err != nil {
+			return err
+		}
+		if _, err := gitCmd(ctx, tmpDir, "fetch", "--depth=1", "file://"+src, ref.Ref); err != nil {
+			return err
+		}
+		if _, err := gitCmd(ctx, tmpDir, "checkout", "FETCH_HEAD"); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if discardGitDir {
+		err := os.RemoveAll(filepath.Join(tmpDir, ".git"))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	bk, err := ref.Query.Buildkit(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get buildkit client: %w", err)
+	}
+	pbDef, _, err := bk.EngineContainerLocalImport(ctx, platforms.DefaultSpec(), tmpDir, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to import container tarball from engine container filesystem: %w", err)
+	}
```

Comment:
As mentioned by @sipsma on [discord](https://discord.com/channels/707636530424053791/1003718839739105300/1313204645316919317), we should avoid the use of this API entirely - it caches very badly, and is not very fun to maintain.

> Updating/adding an LLB source is a better route. I think we could actually customize or even add new LLB ops (not just sources) now that we have our own worker too (just need to handle calling it in [ResolveOp](https://github.com/sipsma/dagger/blob/064c414e4671bbf797bf748b3782e1096fb9f150/engine/buildkit/worker.go#L119-L119)).
>
> Those options give you access to raw snapshots/mounts so you should be able to do whatever you want